gapis/api/vulkan/minimize_textures.go

```diff
@@ -0,0 +1,185 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vulkan
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/transform"
+)
+
+// minimizeTextures returns a transform that sets the size of textures to 1x1
+func minimizeTextures(ctx context.Context) transform.Transformer {
+	ctx = log.Enter(ctx, "Minimize textures")
+	return transform.Transform("Minimize textures", func(ctx context.Context,
+		id api.CmdID, cmd api.Cmd, out transform.Writer) {
+
+		const newTexWidth = 1
+		const newTexHeight = 1
+
+		s := out.State()
+		a := s.Arena
+		l := s.MemoryLayout
+		cb := CommandBuilder{Thread: cmd.Thread(), Arena: a}
+		cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+		switch cmd := cmd.(type) {
+		case *VkCreateImage:
+			imageCreateInfo := cmd.PCreateInfo().MustRead(ctx, cmd, s, nil)
+			if 0 != (imageCreateInfo.Usage() & VkImageUsageFlags(VkImageUsageFlagBits_VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT)) {
+				out.MutateAndWrite(ctx, id, cmd)
+				break
+			}
+
+			imageCreateInfo.SetExtent(NewVkExtent3D(a, newTexWidth, newTexHeight, imageCreateInfo.Extent().Depth()))
+			imageCreateInfo.SetMipLevels(1)
+
+			imageCreateInfoData := s.AllocDataOrPanic(ctx, imageCreateInfo)
+			defer imageCreateInfoData.Free()
+
+			newCmd := cb.VkCreateImage(
+				cmd.Device(),
+				imageCreateInfoData.Ptr(),
+				cmd.PAllocator(),
+				cmd.PImage(),
+				VkResult_VK_SUCCESS,
+			).AddRead(imageCreateInfoData.Data())
+
+			for _, w := range cmd.Extras().Observations().Writes {
+				newCmd.AddWrite(w.Range, w.ID)
+			}
+			out.MutateAndWrite(ctx, id, newCmd)
+		case *VkCmdCopyBufferToImage:
+			bufferImageCopy := cmd.PRegions().Slice(0, 1, l).MustRead(ctx, cmd, s, nil)[0]
+			imageSubresource := bufferImageCopy.ImageSubresource()
+			imageSubresource.SetMipLevel(0)
+			bufferImageCopy.SetImageSubresource(imageSubresource)
+			bufferImageCopy.SetImageExtent(NewVkExtent3D(a, newTexWidth, newTexHeight, bufferImageCopy.ImageExtent().Depth()))
```

Comment:
I think this is valid but we should double check against the valid usage of VkBufferImageCopy
```

If the calling command’s VkImage parameter is a compressed image, or a single-plane, “_422” image format, imageExtent.width must be a multiple of the compressed texel block width or (imageExtent.width + imageOffset.x) must equal the image subresource width

If the calling command’s VkImage parameter is a compressed image, or a single-plane, “_422” image format, imageExtent.height must be a multiple of the compressed texel block height or (imageExtent.height + imageOffset.y) must equal the image subresource height
```